Add tutorial (#421)
pester/docs: tutorial/4-mocking/2-your-first-mock.mdx

---
id: your-first-mock
title: Your first mock
description: Replace a module's internal data source with Mock -ModuleName so tests control their own data instead of depending on a file on disk
---

Your tests currently depend on the contents of `planets.csv`. Add a planet to that file and `Returns all eight planets` starts failing — a test about filtering, broken by a data edit.

A mock replaces a command for the duration of a test. Instead of `Get-PlanetData` reading a file, it returns whatever you tell it to.

## Mocking inside a module

```powershell title="Planetarium/Public/Get-Planet.Mocking.Tests.ps1"
BeforeAll {

>     Import-Module $PSScriptRoot/../Planetarium.psd1 -Force

Author: ```suggestion
    Import-Module "$PSScriptRoot/../Planetarium.psd1" -Force
```

## Why `-ModuleName` is required

This is the detail that costs people the most time.

`Get-Planet` runs inside the module's own scope. When it calls `Get-PlanetData`, PowerShell resolves that name *within the module*, and a mock defined in your test file lives outside it. Without `-ModuleName`, your mock sits somewhere the module will never look, and the real function runs — your test passes or fails for reasons unrelated to the mock you thought you installed.

`-ModuleName Planetarium` injects the mock into the module's scope, where the call actually resolves.

:::warning A mock without `-ModuleName` fails silently
Nothing errors. The mock is simply never used, and the real command runs instead. If a mock appears to have no effect, this is the first thing to check.
:::

Note also what you did *not* have to do: no `InModuleScope` wrapper. `-ModuleName` reaches the private `Get-PlanetData` without dragging your whole test inside the module — which is exactly the preference the [modules guide](../../docs/usage/modules) recommends over `InModuleScope`.

> ## Mocking a built-in command

Author: Move this part to Private/Get-PlanetData.Tests.ps1 ?

> You can mock commands you did not write, including PowerShell's own. Here the target is `Import-Csv`, one layer deeper:

Author: ```suggestion
You can mock commands you did not write, including PowerShell's own. Here the target is `Import-Csv`, one layer deeper. Add this `Describe` block to the same test file:
```

```powershell title="Planetarium/Public/Get-Planet.Mocking.Tests.ps1"
Describe 'Get-PlanetData' {
    It 'Converts the CSV strings into numbers' {
        Mock -ModuleName Planetarium Import-Csv {
            @([PSCustomObject] @{ Name = 'Aiur'; Order = '3'; DistanceFromSunKm = '149597870.7' })
        }

> 
>         InModuleScope Planetarium {
>             $planet = Get-PlanetData
>             $planet.Order | Should-HaveType ([int])
>             $planet.DistanceFromSunKm | Should-HaveType ([double])

Author: ```suggestion
            $planet = Get-PlanetData
            $planet.Name | Should-Be 'Aiur'
            $planet.Order | Should-Be 3
            $planet.Order | Should-HaveType ([int])
            $planet.DistanceFromSunKm | Should-HaveType ([double])
```

At least assert the mocked data.

        }
    }
}
```

This is the test that closes the gap left at the end of the previous page. Feeding in strings — `'3'`, not `3` — and then asserting on the types that come out is what proves the conversion happens. `Should-HaveType` is the assertion that can do it: `Should-Be` would pass either way, because it compares values and lets PowerShell convert.

The mock is what makes the input certain. `Import-Csv` happens to return strings today, so this test would pass without it — but only by accident of the current storage format. Swap the CSV for JSON, which preserves numbers, and an unmocked version of this test would go green without the conversion ever running. Mocking the input pins the test to the behaviour rather than to the file format.

> Delete the `ForEach-Object` block from `Get-PlanetData` and run the suite: this is now the one test that fails.

Author: ```suggestion
Try deleting the `ForEach-Object` block from `Get-PlanetData` and run the suite: this is now the one test that fails. Restore it after, before continuing.
```
